cpp/src/umap/simpl_set_embed/algo.h

```diff
@@ -131,23 +130,163 @@ __device__ __host__ double attractive_grad(double dist_squared,
  * randomly.
  */
 template <typename T, int TPB_X>
-__global__ void optimize_batch_kernel(
+__global__ void optimize_batch_kernel_on_shared_mem(
   T *head_embedding, int head_n, T *tail_embedding, int tail_n, const int *head,
   const int *tail, int nnz, T *epochs_per_sample, int n_vertices,
   bool move_other, T *epochs_per_negative_sample,
   T *epoch_of_next_negative_sample, T *epoch_of_next_sample, double alpha,
   int epoch, double gamma, uint64_t seed, UMAPParams params) {
+  extern __shared__ T embedding_updates[];
+
+  int row = (blockIdx.x * TPB_X) + threadIdx.x;
+  if (row < nnz) {
+    /**
+     * Positive sample stage (attractive forces)
+     */
+
+    if (epoch_of_next_sample[row] <= epoch) {
+      int j = head[row];
+      int k = tail[row];
+
+      T *current = head_embedding + (j * params.n_components);
+      T *other = tail_embedding + (k * params.n_components);
+      T *current_buffer, *other_buffer;
+
+      if (move_other) {
+        current_buffer =
+          embedding_updates + ((threadIdx.x * 2) * params.n_components);
+        other_buffer =
+          embedding_updates + (((threadIdx.x * 2) + 1) * params.n_components);
+      } else {
+        current_buffer =
+          embedding_updates + (threadIdx.x * params.n_components);
+      }
+
+      // initialization of shared memory
+      for (int d = 0; d < params.n_components; d++) {
```

Comment:
Most of the logic in these kernels is exactly the same. I'm concerned having these different code paths with largely the same logic is going to become a maintenance burden when we need to make updates. The nice thing is, these are largely decomposable so I think different `__device__` functions can be made for the common parts.

Reply:
Ok, I'll try to do that.

Reply:
So, finally I found that templating might also allow us to work on a single kernel. Moreover, I think that the compiler optimizes when it sees `if (true)` or `if (false)` by discarding the unnecessary branch.